card-gallery/expressive-search.js

```diff
@@ -0,0 +1,968 @@
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.collectionFeaturedIssue, this.regexp);
+  }
+}
+
+class RewardTitleCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.rewardATitle, this.regexp) || regexMatch(c.rewardBTitle, this.regexp);
+  }
+}
+
+class RewardFlavorTextCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.rewardAFlavorText, this.regexp) || regexMatch(c.rewardBFlavorText, this.regexp);
+  }
+}
+
+class RewardGameTextCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.rewardAGameText, this.regexp) || regexMatch(c.rewardBGameText, this.regexp);
+  }
+}
+
+class CollectionIssuesCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return listRegexMatch(c.collectionIssues, regexp);
+  }
+}
+
+class QuantityCond {
+  constructor(num, relationship) {
+    this.num = num;
+    this.relationship = relationship;
+  }
+  match(c) {
+    return numberMatch(c.quantity, this.num, this.relationship);
+  }
+}
+
+class SetCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.set, this.regexp);
+  }
+}
+
+class ComplexityCond {
+  constructor(num, relationship) {
+    this.num = num;
+    this.relationship = relationship;
+  }
+  match(c) {
+    return numberMatch(c.complexity, this.num, this.relationship);
+  }
+}
+
+class DifficultyCond {
+  constructor(num, relationship) {
+    this.num = num;
+    this.relationship = relationship;
+  }
+  match(c) {
+    return numberMatch(c.difficulty, this.num, this.relationship);
+  }
+}
+
+class TypeCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.type, this.regexp);
+  }
+}
+
+class KindCond {
+  constructor(regexp) {
+    this.regexp = regexp;
+  }
+  match(c) {
+    return regexMatch(c.kind, this.regexp);
+  }
+}
+
+class HasBackCond {
+  constructor(flag) {
+    this.flag = flag;
+  }
+  match(c) {
+    return xnor(this.flag, c.hasBack);
+  }
+}
+
+/**
+ * Performs an expressive search by decomposing the provided query string into an ANDed set of query conditions.
+ * @returns true if the search was successful. Returns false if any errors occurred parsing / resolving the query, or if the query doesn't use the expressive syntax. 
+ */
+function expressiveSearch(queryString) {
+  queryString = escapeRegex(queryString);
+
+  // Searches without query parameters should be handled in the future, but for now we can just bail and let legacy search handle it. Do a quick check for that.
+  if (!queryString.match(/[:=<>]/i)) {
+    return false;
+  }
+
+  const conds = [];
+  const s = new StringScanner(queryString);
+  try {
+    while (!s.hasTerminated()) {
+      if (s.scan(/[\s,]+/i)) {
+        // pass
+      } else if (s.scan(/(?:g|gametext)\s*[:=]\s*(?:"(.*?)"|([^\s\)]+))/i)) {
```

Comment:
All of the strings should match:
- `token=val`
- `token:val`
- `token="val also spaces are allowed"`
- `token:"val also spaces are allowed"`

The numbers should support `>`, `>=`, `<`, `<=`, `=`, and `:` (treating `;` as `=`).

`back` is a unique token I'm using right now for anything that has a back (chars & events). If I can get OR queries working (that's a big if), I'll delete that, but we might find another use for boolean conditions.

If you have suggestions for other conditions, or other bits of data to parse/track in `Card`, let me know. It's pretty quick to add new stuff now that the foundation has been laid.